server/src/main/java/io/crate/expression/operator/DistinctFrom.java

```diff
@@ -87,4 +105,66 @@ public final Boolean evaluate(TransactionContext txnCtx, NodeContext nodeCtx, In
         }
         return argType.compare(arg1, arg2) != 0;
     }
+
+    @Override
+    public Query toQuery(Function function, LuceneQueryBuilder.Context context) {
+        List<Symbol> args = function.arguments();
+        if (!(args.get(0) instanceof Reference ref && args.get(1) instanceof Literal<?> literal)) {
+            return null;
+        }
+        String storageIdentifier = ref.storageIdent();
+        Object value = literal.value();
+        if (value == null) {
+            if (!ref.isNullable()) {
+                // If the column is NOT NULL, `x IS DISTINCT FROM NULL` is true for all documents
+                return new MatchAllDocsQuery();
+            }
+            return IsNullPredicate.refExistsQuery(ref, context);
+        }
+
+        DataType<?> dataType = ref.valueType();
+        return switch (dataType.id()) {
+            case ObjectType.ID -> {
+                var query = EqOperator.refEqObject(
+                    function,
+                    ref.column(),
+                    (ObjectType) dataType,
+                    (Map<String, Object>) value,
+                    context,
+                    BooleanClause.Occur.MUST_NOT
+                );
+                if (query instanceof GenericFunctionQuery) {
+                    yield query;
+                }
+                if (query instanceof BooleanQuery bq) {
+                    if (bq.clauses().size() != 1) {
+                        yield query;
+                    }
+                    var boolQuery = new BooleanQuery.Builder();
+                    boolQuery.add(bq.clauses().getFirst().getQuery(), BooleanClause.Occur.MUST_NOT);
+                    boolQuery.add(genericFunctionFilter(function, context), BooleanClause.Occur.FILTER);
+                    yield boolQuery.build();
+                }
+                yield query;
+            }
+            case ArrayType.ID -> EqOperator.termsAndGenericFilter(
+                function,
+                storageIdentifier,
+                dataType,
+                (Collection<?>) value,
+                context,
+                ref.hasDocValues(),
+                ref.indexType(),
+                BooleanClause.Occur.MUST_NOT
+            );
+            default -> {
+                var query = EqOperator.fromPrimitive(dataType, storageIdentifier, value, ref.hasDocValues(), ref.indexType());
+                if (query == null) {
+                    yield null;
+                }
+                yield query instanceof GenericFunctionQuery ? query : Queries.not(query);
```

Comment:
This also seems a bit odd, imho `fromPrimitive` should never return queries that have different semantics that require a check like this.

Reply:
The problem here is the negation, the generic function filter for `IS DISTINCT` must not be negated while a eq-query must. Any suggestion how we could do this in a better way without duplicating the EQ logic?

Reply:
Looking at this in more detail it now makes even less sense to me. `fromPrimitive` never returns a `GenericFunctionQuery` as far as I can tell? It is also not receiving the `function`, so I'm not sure how it could ?

Reply:
Very right, seems like this was some left-over during development. Can be removed, sorry for the noise.